### PR TITLE
[Serve] Refactor to Fix Type Checking Errors

### PR DESCRIPTION
**Pull request by andylizf**
This PR refactors multiple files within the Serve module to address type-checking errors and improve type annotations. Key changes include:

- Adjustments to method signatures and variable types in modules like autoscalers.py and replica_managers.py for better type safety.
- Modifications to ensure consistent use of `aiohttp.ClientTimeout` in load_balancer.py to prevent timeouts.
- Refactoring of methods in serve_state.py to make paths considered as `str` rather than `str | Path` using helper functions.
- Cleaning up and streamlining processes in service.py for managing service states and cleanup tasks.

Tested (run the relevant ones):

- [x] Code formatting: `bash format.sh`
- [ ] Any manual or new tests for this PR (please specify below)
- [ ] All smoke tests: `pytest tests/test_smoke.py` 
- [ ] Relevant individual smoke tests: `pytest tests/test_smoke.py::test_fill_in_the_name` 
- [ ] Backward compatibility tests: `conda deactivate; bash -i tests/backward_compatibility_tests.sh`

### CODE REVIEW

File: sky/serve/service.py [hunk 1]
```diff
@@ -116,15 +116,17 @@ def _cleanup(service_name: str) -> bool:
...
-    for version in versions:
+
+    def cleanup_version_storage(version: int) -> bool:
         task_yaml: str = serve_utils.generate_task_yaml_file_name(
             service_name, version)
         logger.info(f'Cleaning up storage for version {version}, '
                     f'task_yaml: {task_yaml}')
-        success = success and cleanup_storage(task_yaml)
-    if not success:
+        return cleanup_storage(task_yaml)
```
Comment (cblmemo) [inline sky/serve/service.py:125]: This cleanup storage should only be execute once

File: sky/serve/service.py [hunk 2]
```diff
@@ -236,13 +239,10 @@ def _get_host():
...
             service_name, serve_state.ServiceStatus.SHUTTING_DOWN)
     finally:
-        process_to_kill: List[multiprocessing.Process] = []
-        if load_balancer_process is not None:
-            process_to_kill.append(load_balancer_process)
-        if controller_process is not None:
-            process_to_kill.append(controller_process)
         # Kill load balancer process first since it will raise errors if failed
         # to connect to the controller. Then the controller process.
+        process_to_kill = filter(None,
```
Comment (cblmemo) [inline sky/serve/service.py:244]:
```suggestion
        process_to_kill = filter(lambda proc: proc is not None,
```
for better readability

**Review by cblmemo — COMMENTED** [review]
@andylizf Thanks for the refactor! This mostly looks good to me. Left some nits

File: sky/serve/service.py (hunk 1 above)
Comment (andylizf) [reply sky/serve/service.py:125]: What do you mean by “execute once”. IIUC, the previous code cleans up storage for each versions, until encounters a failure with a version.

Comment (cblmemo) [reply sky/serve/service.py:125]: oh nvm, i misread the code. please ignore this one

File: sky/serve/service.py (hunk 2 above)
Comment (cblmemo) [reply sky/serve/service.py:244]: Actually, after digging deep into the function, setting the first argument to `None` will not only filter out `None` but all false variables, which is not the suggested way to check `None` (always use `if xx is (not) None` to check this, reference [here](https://google.github.io/styleguide/pyguide.html#2144-decision)). Can we try sth like `process_to_kill=[proc for proc in [lbproc, controllerproc] if proc is not None]`?

image

Comment (andylizf) [reply sky/serve/service.py:244]: Good idea!

File: sky/serve/service.py [hunk 3]
```diff
@@ -236,13 +239,12 @@ def _get_host():
...
-        if controller_process is not None:
-            process_to_kill.append(controller_process)
         # Kill load balancer process first since it will raise errors if failed
         # to connect to the controller. Then the controller process.
+        process_to_kill = [
+            proc for proc in [load_balancer_process, controller_process]
+            if proc is not None
+        ]
         subprocess_utils.kill_children_processes(
             [process.pid for process in process_to_kill], force=True)
```
Comment (cblmemo) [inline sky/serve/service.py:249]:
```suggestion
        pids_to_kill = [
            proc.pid for proc in [load_balancer_process, controller_process]
            if proc is not None
        ]
        subprocess_utils.kill_children_processes(
            pids_to_kill, force=True)
```
nit: lets merge the two list comprehension together ;)

**Review by cblmemo — APPROVED** [review]
Thanks for the quick fix @andylizf ! Left one last nit and after that it should be ready to go 🙌

File: sky/serve/service.py (hunk 3 above)
Comment (andylizf) [reply sky/serve/service.py:249]: Thanks for the suggestion! However, we still need the `process_to_kill` list for the `process.join()` call in the following lines, so I can't merge the list comprehensions.

Comment (cblmemo) [reply sky/serve/service.py:249]: Good point!

**Pull request merged**